server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -87,6 +86,64 @@ public Builder add(Mapper.Builder builder) {
             return this;
         }
 
+        Builder addMappers(Map<String, Mapper> mappers) {
+            mappers.forEach((name, mapper) -> mappersBuilders.add(new Mapper.Builder(name) {
+                @Override
+                public Mapper build(MapperBuilderContext context) {
+                    return mapper;
+                }
+            }));
+            return this;
+        }
+
+        /**
+         * Adds a dynamically created Mapper to this builder.
+         *
+         * @param name      the name of the Mapper, including object prefixes
+         * @param prefix    the object prefix of this mapper
+         * @param mapper    the mapper to add
+         * @param context   the DocumentParserContext in which the mapper has been built
+         */
+        public void addDynamic(String name, String prefix, Mapper mapper, DocumentParserContext context) {
+            // If the mapper to add has no dots and is therefore
+            // a leaf mapper, we just add it here
+            if (name.contains(".") == false) {
+                mappersBuilders.add(new Mapper.Builder(name) {
+                    @Override
+                    public Mapper build(MapperBuilderContext context) {
+                        return mapper;
+                    }
+                });
+            }
+            // otherwise we strip off the first object path of the mapper name, load or create
+            // the relevant object mapper, and then recurse down into it, passing the remainder
+            // of the mapper name. So for a mapper 'foo.bar.baz', we locate 'foo' and then
+            // call addDynamic on it with the name 'bar.baz'.
+            else {
+                int firstDotIndex = name.indexOf(".");
+                String childName = name.substring(0, firstDotIndex);
+                String fullChildName = prefix == null ? childName : prefix + "." + childName;
+                ObjectMapper.Builder childBuilder = findChild(childName, fullChildName, context);
+                childBuilder.addDynamic(name.substring(firstDotIndex + 1), fullChildName, mapper, context);
+                mappersBuilders.add(childBuilder);
+            }
+        }
+
+        private ObjectMapper.Builder findChild(String childName, String fullChildName, DocumentParserContext context) {
+            // does the child mapper already exist? if so, use that
+            ObjectMapper child = context.mappingLookup().objectMappers().get(fullChildName);
+            if (child != null) {
+                return child.newBuilder();
+            }
+            // has the child mapper been added as a dynamic update already?
+            child = context.getObjectMapper(fullChildName);
```

Comment:
Could this be named `getDynamicObjectBuilder`?

Reply:
I've renamed it to `getDynamicObjectMapper` - it doesn't return a builder, it returns a mapper that we then call `newBuilder()` on.